[RFC][Codegen] Add custom namespace support

larryliu0820 opened the pull request:
Summary:
Adding a feature to allow user to specify namespaces for operator and kernels.

# Feature
There's a feature request to allow DSL to:
1. take in an operator namespace other than `aten`.
2. take in a kernel that is in a different namespace than `at::native`.


For both features, we only allow user to have a single layer of namespace for the sake of simplicity. If user specify `custom::function` as kernel, the codegen will depend on `custom::native::function` where `native` is hardcoded.

# Proposal

For feature 1, add a `namespace` attribute to data class `NativeFunction`. The namespace will be extract out by matching pattern "::" on the `func` variable. For `NativeFunctionsGroup` there's an assumption that all variants (function, inplace, out) will have the same namespace. By default (if not specified) the namespace will be "aten".

For feature 2, add a `namespace` attribute to `BackendMetadata` class, similarly match pattern "::" on the kernel field. Remove the `cpp_namespace` field from `register_dispatch_key` data class. By default (if not specified) the namespace for a kernel would be "at::native".

Test Plan:
Example yaml entries:
```
- func: custom::gelu.out(Tensor self, *, str approximate='none', Tensor(a!) out) -> Tensor(a!)
  structured: True
  structured_inherits: TensorIteratorBase
  device_check: NoCheck   # TensorIterator
  python_module: nn
  dispatch:
    CPU: custom::gelu_out_cpu
    CUDA: custom::gelu_out_cuda
    MPS: custom::gelu_out_mps

- func: custom::gelu_(Tensor(a!) self, *, str approximate='none') -> Tensor(a!)
  structured_delegate: gelu.out
  device_check: NoCheck   # TensorIterator
  python_module: nn
  dispatch:
    NestedTensorCPU, NestedTensorCUDA: custom::NestedTensor_gelu_

- func: custom::gelu(Tensor self, *, str approximate='none') -> Tensor
  structured_delegate: gelu.out
  device_check: NoCheck   # TensorIterator
  python_module: nn
  dispatch:
    MkldnnCPU: custom::mkldnn_gelu
    QuantizedCPU: custom::gelu_quantized_cpu
    NestedTensorCPU, NestedTensorCUDA: custom::NestedTensor_gelu
```

see generated code:

`RegisterCPU.cpp`:
```
TORCH_LIBRARY_IMPL(aten, CPU, m) {
  ...
}
TORCH_LIBRARY_IMPL(custom, CPU, m) {
    m.impl("gelu", TORCH_FN(wrapper_gelu));
    m.impl("gelu.out", TORCH_FN(wrapper_gelu_out_out));
    m.impl("gelu_", TORCH_FN(wrapper_gelu_));
};
```
```
struct structured_gelu_out_cpu_inplace final : public custom::native::structured_gelu_out_cpu {
    structured_gelu_out_cpu_inplace(Tensor& self) : outputs_{std::ref(self)} {}

    void set_output_strided(
        int64_t output_idx, IntArrayRef sizes, IntArrayRef strides,
        TensorOptions options, DimnameList names
    ) override {

        const auto& out = outputs_[output_idx].get();
        check_inplace(out, sizes, options);

        auto maybe_proxy = maybe_create_proxy(out, sizes, strides, options);
        if (C10_UNLIKELY(maybe_proxy.has_value())) {
            proxy_outputs_[output_idx] = c10::ExclusivelyOwned<Tensor>(std::move(maybe_proxy).value());
        }

        if (!names.empty()) {
          namedinference::propagate_names(outputs_[output_idx], names);
        }
        // super must happen after, so that downstream can use maybe_get_output
        // to retrieve the output
        custom::native::structured_gelu_out_cpu::set_output_raw_strided(output_idx, sizes, strides, options, names);
    }

    void set_output_raw_strided(
        int64_t output_idx, IntArrayRef sizes, IntArrayRef strides,
        TensorOptions options, DimnameList names
    ) override {

        const auto& out = outputs_[output_idx].get();
        check_inplace(out, sizes, options);

        if (!names.empty()) {
          namedinference::propagate_names(outputs_[output_idx], names);
        }
        // super must happen after, so that downstream can use maybe_get_output
        // to retrieve the output
        custom::native::structured_gelu_out_cpu::set_output_raw_strided(output_idx, sizes, strides, options, names);
    }

    const Tensor& maybe_get_output(int64_t output_idx) override {
      return proxy_outputs_[output_idx].has_value() ? **proxy_outputs_[output_idx] : outputs_[output_idx].get();

    }
    std::array<std::reference_wrapper<Tensor>, 1> outputs_;
    std::array<c10::optional<c10::ExclusivelyOwned<Tensor>>, 1> proxy_outputs_;
};
```

`RegisterSchema.cpp`
```
TORCH_LIBRARY(aten, m) {
  ...
}
TORCH_LIBRARY(custom, m) {
    m.def("gelu.out(Tensor self, *, str approximate='none', Tensor(a!) out) -> Tensor(a!)");
    
    m.def("gelu_(Tensor(a!) self, *, str approximate='none') -> Tensor(a!)");
    
    m.def("gelu(Tensor self, *, str approximate='none') -> Tensor");
};
```

Differential Revision: D36558459

ezyang commented:
Namespace customization in dispatch seems like it would be a pain. Maybe simplify by only allowing one level of hierarchy

ezyang commented on aten/src/ATen/templates/RegisterSchema.cpp:
```diff
@@ -4,10 +4,9 @@
   // Implementations located in torch/csrc/jit/runtime/register_distributed_ops.cpp
   m.def("get_gradients(int context_id) -> Dict(Tensor, Tensor)");
 }
+${schema_registrations}
```
?

larryliu0820 replied:
Need to bring this out since the namespace is a variable now and I don't want to lose the registration of "get_gradients".

ezyang commented on torchgen/model.py:
```diff
@@ -440,7 +446,13 @@ def from_yaml(
         assert isinstance(funcs, str), f"not a str: {funcs}"
-        func = FunctionSchema.parse(funcs)
+
+        namespaced_funcs = funcs.split("::")
```
We don't support arbitrary levels of nesting namespaces so you can pass 2 to split

ezyang commented on torchgen/model.py:
```diff
@@ -440,7 +446,13 @@ def from_yaml(
+        if len(namespaced_funcs) == 1:
+            namespace = "aten"
+        else:
+            namespace = "::".join(namespaced_funcs[:-1])
```
and so this would never be exercised

ezyang commented on torchgen/model.py:
```diff
@@ -581,7 +595,7 @@ def from_yaml(
         elif not structured and structured_delegate is None:
             dispatch[DispatchKey.CompositeImplicitAutograd] = BackendMetadata(
-                cpp.name(func), structured=False
+                cpp.name(func), structured=False, cpp_namespace=""
```
How come these aren't defaulted

ezyang commented on torchgen/gen.py:
```diff
@@ -2063,12 +2070,33 @@ def gen_backend_select() -> Dict[str, List[str]]:
+TORCH_LIBRARY($ns, m) {
+    $schema_registrations_body
+};"""
+        )
```
Modern idiomatic codegen uses f-strings

ezyang commented:
I guess I'm kind of waiting for CI to be more green before taking another look. If you need help debugging failures holler

bdhirsh commented on torchgen/gen_backend_stubs.py:
```diff
@@ -123,7 +123,7 @@ def create_backend_index(
             kernel_name = dispatcher.name(native_functions_map[op_name].func)
             # TODO: allow structured external backends later.
-            m = BackendMetadata(kernel=kernel_name, structured=False)
+            m = BackendMetadata(kernel=kernel_name, structured=False, cpp_namespace="at::native")
```
What was the original default? empty string?

bdhirsh commented on torchgen/gen.py:
```diff
@@ -2076,12 +2086,38 @@ def gen_backend_select() -> Dict[str, List[str]]:
     schema_selector = selector
     if force_schema_registration:
         schema_selector = SelectiveBuilder.get_nop_selector()
+    schema_static_template = CodeTemplate(
```
nit: we try to stop using `CodeTemplate` in the codegen, and use `f-strings` instead (the existing `CodeTemplate` usages should just be removed at some point)

bdhirsh commented on torchgen/gen.py:
```diff
@@ -2076,12 +2086,38 @@ def gen_backend_select() -> Dict[str, List[str]]:
+            mapMaybe(RegisterSchema(schema_selector), functions)
+        )
+        if namespace == "aten":
+            aten_schema_registrations = schema_registrations_body
```
I think this deserves a comment - you're treating "aten` specially, because you don't want to generate the `TORCH_LIBRARY` call for it (because the C++ template file already has it)

bdhirsh commented on torchgen/gen.py:
```diff
@@ -2076,12 +2086,38 @@ def gen_backend_select() -> Dict[str, List[str]]:
+        if namespace == "aten":
+            aten_schema_registrations = schema_registrations_body
+        else:
+            schema_registrations += schema_static_template.substitute(
```
It looks like this gives the yaml author total freedom to add multiple different namespaces if they want to. Do we actually need that today? If we do then I think this is fine, but if you only really need "aten + one other custom namespace" for mobile today, then I think we should be keep it simpler and limit the codegen to that.

bdhirsh commented on torchgen/model.py:
```diff
@@ -489,7 +501,11 @@ def from_yaml(
             structured_delegate_s, str
-        ), f"not a str: {structured_delegate}"
+        ), f"not a str: {structured_delegate_s}"
+        assert structured_delegate_s is None or "::" not in structured_delegate_s, (
```
oh if we're gonna add this check, we should probably do the same for `structured_inherits`

bdhirsh replied:
I don't think i see this check?

larryliu0820 replied:
Oops yeah I missed it

bdhirsh commented on torchgen/model.py:
```diff
@@ -552,13 +568,19 @@ def from_yaml(
                         and is_structured_dispatch_key(dispatch_key),
+                        cpp_namespace=f"{tokens[0]}::native"
+                        if len(tokens) > 1
+                        else "at::native",
```
Tiny nit, but maybe it would be clearer that "`at` is the default namespace" if you write this as `namespace = tokens[0] if len(tokens) == 1 else 'at'; cpp_namespace = f"{namespace}::native"`, to make it clear what the default is?

bdhirsh commented on torchgen/model.py:
```diff
@@ -581,7 +603,7 @@ def from_yaml(
         elif not structured and structured_delegate is None:
             dispatch[DispatchKey.CompositeImplicitAutograd] = BackendMetadata(
-                cpp.name(func), structured=False
+                cpp.name(func), structured=False, cpp_namespace="at::native"
```
Actually, the logic above says that "all kernels should live in some {foo}::native" namespace, but it's hard to enforce that invariant because you're obligated to write out the full cpp namespace every place that you construct a `BackendMetadata` object.

I think it would be less error prone if you had the `BackendMetadata` only take in the first part of the namespace (`at`), and have the __init__ function take care of appending a `::native` to it. And maybe add a `DEFAULT_NAMESPACE = "at"` global somewhere in the codegen that all of the call-sites have to use, to make it clear that it should be consistent everywhere.

bdhirsh replied:
this one too, I think?

larryliu0820 replied:
Yeah we can't do this since lazy tensor codegen is having a namespace `torch::lazy`. We can't append `::native` to that.

bdhirsh commented:
you'll also need to run the lint (`lintrunner -a`) in OSS in order to make CI happy

bdhirsh commented:
Just the last 2 comments. And you'll also need to run the lint in OSS to fix the lint tests. More instructions on it here: https://fb.workplace.com/groups/pytorch.dev/permalink/1101083637136782/

Otherwise I think it looks good!

iseeyuan commented:
> For both features, we only allow user to have a single layer of namespace for the sake of simplicity. If user specify custom::function as kernel, the codegen will depend on custom::native::function where native is hardcoded.

I don't fully follow it. Does "single layer of namespace" mean we only allow "custom::function", but not "custom::Jarvis::function"? I'm wondering if it's a good requirement on the kernel side. That would limit the user kernel implementation, where it's normal to have multiple layers of namespace in C++. 

Also any reason we should insert "native" in "custom::function"?

bdhirsh reviewed (approved):
codegen looks ok to me

larryliu0820 committed:
[executorch] ATen integration: add namespace support (#78015)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/78015

Adding a feature to allow user to specify namespaces for operator and kernels.

# Feature
There's a feature request to allow DSL to:
1. take in an operator namespace other than `aten`.
2. take in a kernel that is in a different namespace than `at::native`.

For both features, we only allow user to have a single layer of namespace for the sake of simplicity. If user specify `custom::function` as kernel, the codegen will depend on `custom::native::function` where `native` is hardcoded.

# Proposal

For feature 1, add a `namespace` attribute to data class `NativeFunction` as well as `NativeFunctionsGroup`. The namespace will be extract out by matching pattern "::" on the `func` variable. For `NativeFunctionsGroup` there's an assumption that all variants (function, inplace, out) will have the same namespace. By default (if not specified) the namespace will be "aten".

For feature 2, add a `namespace` attribute to `BackendMetadata` class, similarly match pattern "::" on the kernel field. By default (if not specified) the namespace for a kernel would be "at::native".

Test Plan:
Example yaml entries:
```
- func: custom::gelu.out(Tensor self, *, str approximate='none', Tensor(a!) out) -> Tensor(a!)
  structured: True
  structured_inherits: TensorIteratorBase
  device_check: NoCheck   # TensorIterator
  python_module: nn
  dispatch:
    CPU: custom::gelu_out_cpu
    CUDA: custom::gelu_out_cuda
    MPS: custom::gelu_out_mps

- func: custom::gelu_(Tensor(a!) self, *, str approximate='none') -> Tensor(a!)
  structured_delegate: gelu.out
  device_check: NoCheck   # TensorIterator
  python_module: nn
  dispatch:
    NestedTensorCPU, NestedTensorCUDA: custom::NestedTensor_gelu_

- func: custom::gelu(Tensor self, *, str approximate='none') -> Tensor
  structured_delegate: gelu.out
  device_check: NoCheck   # TensorIterator
  python_module: nn
  dispatch:
    MkldnnCPU: custom::mkldnn_gelu
    QuantizedCPU: custom::gelu_quantized_cpu
    NestedTensorCPU, NestedTensorCUDA: custom::NestedTensor_gelu
```

see generated code:

`RegisterCPU.cpp`:

{F734056529}
```
struct structured_gelu_out_cpu_inplace final : public custom::native::structured_gelu_out_cpu {
    structured_gelu_out_cpu_inplace(Tensor& self) : outputs_{std::ref(self)} {}

    void set_output_strided(
        int64_t output_idx, IntArrayRef sizes, IntArrayRef strides,
        TensorOptions options, DimnameList names
    ) override {

        const auto& out = outputs_[output_idx].get();
        check_inplace(out, sizes, options);

        auto maybe_proxy = maybe_create_proxy(out, sizes, strides, options);
        if (C10_UNLIKELY(maybe_proxy.has_value())) {
            proxy_outputs_[output_idx] = c10::ExclusivelyOwned<Tensor>(std::move(maybe_proxy).value());
        }

        if (!names.empty()) {
          namedinference::propagate_names(outputs_[output_idx], names);
        }
        // super must happen after, so that downstream can use maybe_get_output
        // to retrieve the output
        custom::native::structured_gelu_out_cpu::set_output_raw_strided(output_idx, sizes, strides, options, names);
    }

    void set_output_raw_strided(
        int64_t output_idx, IntArrayRef sizes, IntArrayRef strides,
        TensorOptions options, DimnameList names
    ) override {

        const auto& out = outputs_[output_idx].get();
        check_inplace(out, sizes, options);

        if (!names.empty()) {
          namedinference::propagate_names(outputs_[output_idx], names);
        }
        // super must happen after, so that downstream can use maybe_get_output
        // to retrieve the output
        custom::native::structured_gelu_out_cpu::set_output_raw_strided(output_idx, sizes, strides, options, names);
    }

    const Tensor& maybe_get_output(int64_t output_idx) override {
      return proxy_outputs_[output_idx].has_value() ? **proxy_outputs_[output_idx] : outputs_[output_idx].get();

    }
    std::array<std::reference_wrapper<Tensor>, 1> outputs_;
    std::array<c10::optional<c10::ExclusivelyOwned<Tensor>>, 1> proxy_outputs_;
};
```

`RegisterSchema.cpp`

{F734056640}

Reviewed By: bdhirsh

Differential Revision: D36558459

fbshipit-source-id: f1282487b08746fa2158ca0106ac7b74a2402899
